src/providers/hana/qgshanaconnectionstringbuilder.cpp

```diff
@@ -52,42 +53,75 @@ QgsHanaConnectionStringBuilder::QgsHanaConnectionStringBuilder( const QgsDataSou
 
 QString QgsHanaConnectionStringBuilder::toString() const
 {
-  // For more details on how to escape special characters in passwords,
-  // see https://stackoverflow.com/questions/55150362/maybe-illegal-character-in-odbc-sql-server-connection-string-pwd
-  const QString pwd = QString( mPassword ).replace( '}', QLatin1String( "}}" ) );
-  QString ret = !mDsn.isEmpty()
-                ? QStringLiteral( "DSN=%1;UID={%2};PWD={%3};CHAR_AS_UTF8=1" ).arg(
-                  mDsn, mUserName, pwd )
-                : QStringLiteral( "DRIVER={%1};SERVERNODE=%2:%3;DATABASENAME=%4;UID={%5};PWD={%6};CHAR_AS_UTF8=1" ).arg(
-                  mDriver, mHost, mPort, mDatabase, mUserName, pwd );
-  if ( !mSchema.isEmpty() )
-    ret += QStringLiteral( ";CURRENTSCHEMA=" ) + mSchema;
+  QStringList props;
+  auto addProperty = [&props]( const QString & name, const QString & value )
+  {
+    // See notes for constructing connection string for HANA
+    // https://help.sap.com/docs/SAP_HANA_CLIENT/f1b440ded6144a54ada97ff95dac7adf/7cab593774474f2f8db335710b2f5c50.html
+
+    if ( value.isEmpty() )
+      return;
+
+    bool hasSpecialChars = false;
+    for ( const QChar ch : value )
+    {
+      if ( QStringLiteral( "[]{}(),;?*=!@" ).indexOf( ch ) != -1 )
```

Comment:
This is an inefficient way to check this -- better to use a regular expression match instead

Reply:
Done in https://github.com/qgis/QGIS/pull/50475/commits/e1a70c739e1f6e2ab092c3a1fceb779da7a46c73.